main.tf

```diff
@@ -1,11 +1,39 @@
+data "aws_iam_policy_document" "this" {
+  # dont create this if no services or iam arns are specified
+  count = length(concat(var.allow_publish_aws_services, var.allow_publish_iam_arns)) > 0 ? 1 : 0
+  statement {
+    effect    = "Allow"
+    actions   = ["sns:Publish"]
+    resources = ["*"]
+
+    dynamic principals {
+      for_each = length(var.allow_publish_aws_services) > 0 ? ["_enable"] : []
+      content {
+        type        = "Service"
+        identifiers = var.allow_publish_aws_services
+      }
+    }
+
+    dynamic principals {
+      for_each = length(var.allow_publish_iam_arns) > 0 ? ["_enable"] : []
+      content {
+        type        = "AWS"
+        identifiers = var.allow_publish_iam_arns
+      }
+    }
+  }
+}
+
+
+
 resource "aws_sns_topic" "this" {
   count = var.create_sns_topic ? 1 : 0
 
   name        = var.name
   name_prefix = var.name_prefix
 
   display_name                             = var.display_name
-  policy                                   = var.policy
+  policy                                   = length(var.policy) > 0 ? var.policy : try(data.aws_iam_policy_document.this[0].json, null)
```

Comment:
I think this is shorter - `policy = try(data.aws_iam_policy_document.this[0].json, var.policy)`

It requires to update `count` in `data "aws_iam_policy_document" "this"`.

Reply:
Roger that, done!